lib/AST/LookupVisibleDecls.cpp

```diff
@@ -473,19 +473,86 @@ lookupVisibleMemberDeclsImpl(Type BaseTy, VisibleDeclConsumer &Consumer,
                              GenericSignatureBuilder *GSB,
                              VisitedSet &Visited);
 
-static void lookupVisibleProtocolMemberDecls(
-    Type BaseTy, ProtocolType *PT, VisibleDeclConsumer &Consumer,
-    const DeclContext *CurrDC, LookupState LS, DeclVisibilityKind Reason,
-                                             LazyResolver *TypeResolver, GenericSignatureBuilder *GSB,
-    VisitedSet &Visited) {
+class RestateFilteringConsumer : public VisibleDeclConsumer {
+  VisibleDeclConsumer &parentConsumer;
+  const Type baseTy;
+  const DeclContext *DC;
+  LazyResolver *resolver;
+  llvm::DenseSet<DeclName> foundVars;
+  llvm::DenseSet<std::pair<DeclName, CanType>> foundFuncs;
+  void validate() {
+    assert(DC && baseTy && !baseTy->hasLValueType());
+  }
+public:
+  RestateFilteringConsumer(VisibleDeclConsumer &parentConsumer,
+                           Type baseTy, const DeclContext *DC,
+                           LazyResolver *resolver)
+  : parentConsumer(parentConsumer), baseTy(baseTy), DC(DC), resolver(resolver) {
+    validate();
+  }
+
+  void foundDecl(ValueDecl *VD, DeclVisibilityKind Reason) override {
+    assert(VD);
+    // If this isn't a protocol context, hand over the decl
+    // to the parent consumer.
+    if (!isa<ProtocolDecl>(VD->getDeclContext())) {
+      parentConsumer.foundDecl(VD, Reason);
+      return;
+    }
+    if (resolver)
+      resolver->resolveDeclSignature(VD);
+
+    if (!VD->hasInterfaceType()) {
+      parentConsumer.foundDecl(VD, Reason);
+      return;
+    }
+    auto AFT = VD->getInterfaceType()->getAs<AnyFunctionType>();
+    if (!AFT) {
+      if (foundVars.insert(VD->getFullName()).second)
+        parentConsumer.foundDecl(VD, Reason);
+      return;
+    }
+    // Preserve the generic signature if this is a subscript, which are uncurried,
+    // or if we have genenic params other than Self. Otherwise, strip if off
+    // and use the resultType of the curried function type.
+    // When preserving the generic signature, we remove the requirement
+    // from Self to make sure it doesn't prevent us from recognizing restatements.
+    CanType type;
+    // This can't be null since we are dealing with method or
+    // subscript requirements, where Self is an implicit generic param.
+    auto sig = AFT->getOptGenericSignature();
+    auto params = sig->getGenericParams();
+    if (params.size() == 1 && !isa<SubscriptDecl>(VD)) {
+      type = AFT->getResult()->getCanonicalType();
+    } else {
+      auto newReqs = sig->getRequirements().drop_front();
```

Comment:
Please don’t make assumptions about the order of requirements in the generic signature like this. There must be a better way to accomplish what you’re trying to do. Perhaps you want to drop any conformance requirement with Self on the left hand side instead of assuming it’s the first one. Also this kind of thing should be factored out into a new method; the code here is getting a bit too long.

Reply:
It will be longer than 30 lines once you do the right thing here, and it's more about single responsibility per method than raw length.

Reply:
I see, you mean the type-computing logic.